src/com/amazon/corretto/crypto/provider/HpkeCipher.java

```diff
@@ -0,0 +1,309 @@
+    }
+    if ((opmode == Cipher.UNWRAP_MODE) && !(key instanceof EvpHpkePrivateKey)) {
+      throw new IllegalStateException("Need PrivateKey to unwrap");
+    }
+    synchronized (lock_) {
+      params_ = (HpkeParameterSpec) params;
+      javaCipherMode_ = opmode;
+      key_ = opmode == Cipher.WRAP_MODE ? (EvpHpkePublicKey) key : (EvpHpkePrivateKey) key;
+    }
+  }
+
+  /**
+   * Performs HPKE single-shot encryption, as specified in Section 6.1 of RFC 9180.
+   *
+   * @return concatenation of KEM encapsulated key and encrypted ciphertext
+   */
+  @Override
+  protected byte[] engineWrap(Key key) throws IllegalBlockSizeException, InvalidKeyException {
+    if (javaCipherMode_ != Cipher.WRAP_MODE) {
+      throw new IllegalStateException("Cipher must be in WRAP_MODE");
+    }
+    if ((key_ == null) || !(key_ instanceof EvpHpkePublicKey)) {
+      throw new IllegalStateException("PublicKey should be set before wrapping.");
+    }
+    try {
+      final byte[] encoded = Utils.encodeForWrapping(provider_, key);
+      return engineDoFinal(encoded, 0, encoded.length);
+    } catch (final BadPaddingException e) {
+      throw new InvalidKeyException("Failed to wrap key", e);
+    }
+  }
+
+  /**
+   * Performs HPKE single-shot decryption, as specified in Section 6.1 of RFC 9180.
+   *
+   * @return decrypted plaintext
+   */
+  @Override
+  protected Key engineUnwrap(byte[] wrappedKey, String wrappedKeyAlgorithm, int wrappedKeyType)
+      throws InvalidKeyException, NoSuchAlgorithmException {
+    if (javaCipherMode_ != Cipher.UNWRAP_MODE) {
+      throw new IllegalStateException("Cipher must be in UNWRAP_MODE");
+    }
+    if ((key_ == null) || !(key_ instanceof EvpHpkePrivateKey)) {
+      throw new IllegalStateException("PrivateKey should be set before unwrapping.");
+    }
+    try {
+      final byte[] unwrappedKey = engineDoFinal(wrappedKey, 0, wrappedKey.length);
+      return Utils.buildUnwrappedKey(provider_, unwrappedKey, wrappedKeyAlgorithm, wrappedKeyType);
+    } catch (final BadPaddingException | IllegalBlockSizeException | InvalidKeySpecException e) {
+      throw new InvalidKeyException("Failed to unwrap key", e);
+    }
+  }
+
+  /**
+   * Performs HPKE single-shot encryption or decryption, as specified in Section 6.1 of RFC 9180.
+   *
+   * @return number of bytes written to output
+   */
+  @Override
+  protected int engineDoFinal(
+      byte[] input, int inputOffset, int inputLen, byte[] output, int outputOffset)
+      throws ShortBufferException {
+
+    if (input == output) {
+      // TODO: is it okay if they are non-overlapping, support that case if necessary.
+      throw new IllegalStateException("input and output must be separate arrays");
+    }
+
+    synchronized (lock_) {
+      if (key_ == null) {
+        throw new IllegalStateException("key should be set before finalizing");
+      }
+      if ((javaCipherMode_ != Cipher.WRAP_MODE) && (javaCipherMode_ != Cipher.UNWRAP_MODE)) {
```

Comment:
You're able to encrypt/decrypt arbitrary data by setting the cipher to wrap mode and calling engineDoFinal directly. This seems like unintentional behavior. It can probably be avoided by using a boolean flag to indicate that the corresponding engineWrap method was invoked.

Reply:
Good call out, I addressed this by:
1. moving the implementation from DoFinal to an internal method
2. having wrap and unwrap call that internal method
3. disallowing wrap and unwrap in DoFinal.